File: apple/Sources/UniMusicSync/UniMusicSync.swift
#if canImport(SystemConfiguration)
    import SystemConfiguration
#endif

import Foundation
@_exported import UniFFI

public class UniMusicSync {
    public let irohManager: IrohManager

    public init(_ path: String) async throws {
        let irohFactory = IrohFactory()
        irohManager = try await irohFactory.irohManager(path: path)
    }

    public func shutdown() async throws {
        try await irohManager.shutdown()
    }

    public func createNamespace() async throws -> UNamespaceId {
        let namespace = try await irohManager.createNamespace()
        return namespace
    }

    public func deleteNamespace(namespace: UNamespaceId) async throws {
        try await irohManager.deleteNamespace(namespace: namespace)
    }

    public func getAuthor() async throws -> String {
        let author = try await irohManager.getAuthor()
        return author
    }

    public func getNodeId() async -> UNodeId {
        let nodeId = await irohManager.getNodeId()
        return nodeId
    }

    public func getKnownNodes() async -> [UNodeId] {
        let knownNodes = await irohManager.getKnownNodes()
        return knownNodes
    }

    public func getFiles(_ namespace: UNamespaceId) async throws -> [UEntry] {
        let files = try await irohManager.getFiles(namespace: namespace)
        return files
    }

    public func writeFile(_ namespace: UNamespaceId, _ path: String, _ data: Data) async throws -> UHash {
        let hash = try await irohManager.writeFile(
            namespace: namespace,
            path: path,
            data: data
        )
        return hash
    }

    public func deleteFile(_ namespace: UNamespaceId, _ path: String) async throws -> UInt32 {
        let deletedFiles = try await irohManager.deleteFile(namespace: namespace, path: path)
        return deletedFiles
    }

    public func readFile(_ namespace: UNamespaceId, _ path: String) async throws -> Data {
        let data = try await irohManager.readFile(namespace: namespace, path: path)
        return data
    }

    public func readFileHash(_ hash: UHash) async throws -> Data {
        let data = try await irohManager.readFileHash(hash: hash)
        return data
    }

    public func export(_ namespace: UNamespaceId, _ path: String, _ destination: String) async throws {
        try await irohManager.export(namespace: namespace, path: path, destination: destination)
    }

    public func exportHash(_ hash: UHash, _ destination: String) async throws {
        try await irohManager.exportHash(hash: hash, destination: destination)
    }

    public func share(_ namespace: String) async throws -> UDocTicket {
        let ticket = try await irohManager.share(namespace: namespace)
        return ticket
    }

    public func `import`(_ ticket: String) async throws -> UNamespaceId {
        let namespaceId = try await irohManager.import(ticket: ticket)
        return namespaceId
    }

    public func sync(_ namespace: UNamespaceId) async throws {
        try await irohManager.sync(namespace: namespace)
    }

    public func reconnect() async {
        await irohManager.reconnect()
    }
}

File: android/core/src/main/java/app/unimusic/sync/UniMusicSync.kt
package app.unimusic.sync

import uniffi.unimusic_sync.IrohFactory
import uniffi.unimusic_sync.IrohManager
import uniffi.unimusic_sync.UDocTicket
import uniffi.unimusic_sync.UEntry
import uniffi.unimusic_sync.UHash
import uniffi.unimusic_sync.UNamespaceId
import uniffi.unimusic_sync.UNodeId

@Suppress("unused")
class UniMusicSync(private val irohManager: IrohManager) {
  companion object {
    suspend fun create(path: String): UniMusicSync {
      val irohFactory = IrohFactory()
      val irohManager = irohFactory.irohManager(path)
      val instance = UniMusicSync(irohManager)
      return instance
    }
  }

  // This allows to attach a context, which results in much cleaner stack traces
  private inline fun <T> handleException(block: () -> T): T {
    return try {
      block()
    } catch (exception: Exception) {
      throw RuntimeException("Exception in handleException block", exception)
    }
  }

  fun close() {
    irohManager.close()
  }

  suspend fun shutdown() {
    handleException { irohManager.shutdown() }
  }

  suspend fun createNamespace(): UNamespaceId {
    val namespace = handleException { irohManager.createNamespace() }
    return namespace
  }

  suspend fun deleteNamespace(namespace: UNamespaceId) {
    handleException { irohManager.deleteNamespace(namespace) }
  }

  suspend fun getAuthor(): String {
    val author = handleException { irohManager.getAuthor() }
    return author
  }

  suspend fun getNodeId(): UNodeId {
    val nodeId = irohManager.getNodeId()
    return nodeId
  }

  suspend fun getKnownNodes(): List<UNodeId> {
    val knownNodes = handleException { irohManager.getKnownNodes() }
    return knownNodes
  }

  suspend fun getFiles(namespace: UNamespaceId): List<UEntry> {
    val files = handleException { irohManager.getFiles(namespace) }
    return files
  }

  suspend fun writeFile(namespace: UNamespaceId, path: String, data: ByteArray): UHash {
    val hash = handleException { irohManager.writeFile(namespace, path, data) }
    return hash
  }

  suspend fun deleteFile(namespace: UNamespaceId, path: String): UHash {
    val tombstoneHash = handleException { irohManager.deleteFile(namespace, path) }
    return tombstoneHash
  }

  suspend fun readFile(namespace: UNamespaceId, path: String): ByteArray {
    val data = handleException { irohManager.readFile(namespace, path) }
    return data
  }

  suspend fun readFileHash(hash: UHash): ByteArray {
    val data = handleException { irohManager.readFileHash(hash) }
    return data
  }

  suspend fun export(namespace: UNamespaceId, path: String, destination: String) {
    handleException { irohManager.export(namespace, path, destination) }
  }

  suspend fun exportHash(hash: UHash, destination: String) {
    handleException { irohManager.exportHash(hash, destination) }
  }

  suspend fun share(namespace: UNamespaceId): UDocTicket {
    val ticket = handleException { irohManager.share(namespace) }
    return ticket
  }

  suspend fun import(ticket: String): UNamespaceId {
    val namespace = handleException { irohManager.import(ticket) }
    return namespace
  }

  suspend fun sync(namespace: UNamespaceId) {
    handleException { irohManager.sync(namespace) }
  }

  suspend fun reconnect() {
    handleException { irohManager.reconnect() }
  }
}
